### PR TITLE
Refactor `getValue<T>` code to reduce duplication

### PR DESCRIPTION
**Pull request by andygrove**
This is the first step in cleaning up some of the `PyExpr` code.

Changes in this PR:

- Introduce a new `DaskPlannerError` that we can use in non-Pythonic Rust methods instead of using DataFusion's error types
- Add `PyExpr::get_scalar_value`
- Refactor all `PyExpr::<type>_value` methods to call `PyExpr::get_scalar_value`
- Adds a few Rust unit tests

I can make this code much more concise with a macro, but I didn't want to go that far since we haven't discussed whether we want to start using macros in this project. I will follow up with a separate PR where we can have that discussion.

### CODE REVIEW

File: dask_planner/src/expression.rs [hunk 1]
```diff
@@ -821,10 +778,27 @@ impl PyExpr {
...
+            _ => Err(DaskPlannerError::Internal(
+                "get_scalar_value() called on non-literal expression".to_string(),
+            )),
+        }
+    }
+}
+
+fn unexpected_literal_value() -> PyErr {
+    DaskPlannerError::Internal("getValue<T>() - Unexpected value".to_string()).into()
+}
```
Comment (charlesbluca) [inline dask_planner/src/expression.rs:796]: Would it make sense to display the invalid scalar type received as part of this error message?

File: dask_planner/src/expression.rs [hunk 2]
```diff
@@ -557,214 +557,156 @@ impl PyExpr {
...
-            Expr::Literal(scalar_value) => match scalar_value {
-                ScalarValue::Float32(iv) => Ok(*iv),
-                _ => Err(py_type_err("getValue<T>() - Unexpected value")),
-            },
-            _ => Err(py_type_err("getValue<T>() - Non literal value encountered")),
+    pub fn float_32_value(&self) -> PyResult<Option<f32>> {
+        match self.get_scalar_value()? {
+            ScalarValue::Float32(value) => Ok(*value),
+            other => Err(unexpected_literal_value(other)),
         }
```
Comment (charlesbluca) [inline dask_planner/src/expression.rs:565]: Is it possible/worthwhile to make a helper function (parametrized on the specific `ScalarValue` variant) for this basic match pattern?

Comment (andygrove) [reply dask_planner/src/expression.rs:565]: That would be easy to do with a macro, but there are pros and cons to using macros, so I planned on doing that in a follow-on PR.

Comment (andygrove) [reply dask_planner/src/expression.rs:565]: We could define this macro:

```rust
macro_rules! extract_scalar_value {
    ($self: expr, $variant: ident) => {
        match $self.get_scalar_value()? {
            ScalarValue::$variant(value) => Ok(*value),
            other => Err(unexpected_literal_value(other)),
        }
    }
}
```

and then most of the get_value methods become one-liners:

```rust
    #[pyo3(name = "getFloat32Value")]
    pub fn float_32_value(&self) -> PyResult<Option<f32>> {
        extract_scalar_value!(self, Float32)
    }
```

The main downside is that there is now another language for the maintainers to become familiar with.

Comment (charlesbluca) [reply dask_planner/src/expression.rs:565]: Thanks for the detailed response, in that case I'm okay with pushing this off to a follow-up PR, and we can have a large discussion around the use of macros.

**Review by charlesbluca — APPROVED** [review]

**Pull request merged**